### PR TITLE
proof of concept for streaming messages in nlu training

### PR DESCRIPTION
**Pull request by twerkmeister**
**DRAFT FOR DISCUSSION PURPOSES **:
- proof of concept for #8407 
- The implementation is pretty rough and just to theoretically test some of the bigger risks (such as, could diet be trained from a stream of messages)
- implemented for CountVectorsFeaturizer, Spacy Featurizer, CRF, and DietClassifier
- main training data object (with unfeaturized messages) stays in memory right now, as so many parts of the codebase use it that way
- features can be successively built and stored on disc (right now just pickling a list of features per message, ultimately a safer and more sophisticated structure should be chosen)
- messages can be joined with features on the fly and consumed as a Generator. For the matching I am using a fingerprint of the message that is just based on the following attributes of the message: text, intent, response, action name, action text, and intent response key. I think this way the fingerprint would be invariant across processing steps and across runs.
- keras.Sequence can be used to hide the generator, but there are probably better abstractions available. Possibly also some multiprocessing here to fill a batch queue in a second thread or so.
- Chunking directories are just based on the component name right now, and if some component requests a message stream, the latest chunking dir will be used
- tested with memory profiling on multiwoz, for details see below
- shuffling is implemented by: using smallish chunks and randomizing chunk reading order + reading multiple chunks at once and then shuffling the acumulated again

**notes**
- I think the methods of many components could really benefit from taking a more message centric perspective. So many of components' methods operate on collections of messages. That makes the functions more complex, harder to understand, and harder to reuse. Was able to throw out a lot of code in the `CountVectorFeaturizer` by focusing more on messages
- `RasaModelData` seems a bit like a parallel world to the `TrainingData` class. I think there are many parallel implementations of the same logic on these classes, such as splitting training data according to some labels. I am not sure this intermediate `RasaModelData` class is really necessary. It seems all that's needed is some logic to arrange messages in a certain way and to turn a bunch of them into tensors.

### CODE REVIEW

File: rasa/nlu/classifiers/diet_classifier.py [hunk 1]
```diff
@@ -835,14 +844,13 @@ def train(
...
 
-        data_generator, validation_data_generator = train_utils.create_data_generators(
-            model_data,
-            self.component_config[BATCH_SIZES],
-            self.component_config[EPOCHS],
-            self.component_config[BATCH_STRATEGY],
-            self.component_config[EVAL_NUM_EXAMPLES],
-            self.component_config[RANDOM_SEED],
+        data_generator, validation_data_generator = (
+            MessageStreamDataGenerator(
```
Comment (wochinge) [inline rasa/nlu/classifiers/diet_classifier.py:848]: do you know if the `Caution:` part from [here](https://www.tensorflow.org/guide/data#consuming_python_generators)  applies?

File: rasa/utils/tensorflow/data_generator.py [hunk 2]
```diff
@@ -423,3 +426,66 @@ def _linearly_increasing_batch_size(self) -> int:
...
+            try:
+                messages.append(next(self.current_message_stream))
+            except StopIteration:
+                break
+        data = self.message_to_data_function(messages)
+        return RasaDataGenerator.prepare_batch(data), None
+
+    def on_epoch_end(self) -> None:
+        """Update the data after every epoch."""
+        self.current_message_stream = self.training_data.stream_featurized_messages(
```
Comment (wochinge) [inline rasa/utils/tensorflow/data_generator.py:489]: how would we e.g. shuffle chunks after epoch end?

File: rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py [hunk 3]
```diff
@@ -734,19 +698,24 @@ def train(
...
-                    training_data.training_examples,
-                )
+        self.initialize(training_data, config, **kwargs)
+
+        msg_pipe = training_data.add_features(None, "test2")
+        try:
+            msg = next(msg_pipe)
+            while msg:
+                features = self.msg_to_features(msg)
+                msg = msg_pipe.send(features)
```
Comment (wochinge) [inline rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py:716]: using `send` is neat!

File: rasa/nlu/classifiers/diet_classifier.py (hunk 1 above)
Comment (twerkmeister) [reply rasa/nlu/classifiers/diet_classifier.py:848]: Yeah maybe something like Dataset.from_generator would be the right way to go

File: rasa/utils/tensorflow/data_generator.py (hunk 2 above)
Comment (twerkmeister) [reply rasa/utils/tensorflow/data_generator.py:489]: The TrainingData object could tell the featurereader to read the chunks in a random/different order. And then you still need to assure that the right message gets the right features. Some ID of the message that is also stored in the chunk or so.

File: rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py (hunk 3 above)
Comment (twerkmeister) [reply rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py:716]: Yeah first time for me! It's cool, but easy to get wrong. At first I only stored every second feature set because I both consumed the generator and used send too. Didn't know that send would also make the generator continue and return the next yield immediately.

Comment (wochinge) [reply rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py:716]: It makes sense once you think about it but it's not intuitive 🙈

File: rasa/nlu/classifiers/diet_classifier.py (hunk 1 above)
Comment (wochinge) [reply rasa/nlu/classifiers/diet_classifier.py:848]: so it seems `tf.data` is preferred for non-python preprocessing (== all pre-processing is expressed via TensorFlow functions) while the Keras sequence is preferred if you do a lot of Python preprocessing. So in our case the Keras sequence seems to be preferable.

File: rasa/utils/tensorflow/data_generator.py (hunk 2 above)
Comment (wochinge) [reply rasa/utils/tensorflow/data_generator.py:489]: sweet, thanks for implementing this as well 💯

File: rasa/nlu/extractors/crf_entity_extractor.py [hunk 4]
```diff
@@ -572,14 +582,14 @@ def _train_model(self, df_train: List[List[CRFToken]]) -> None:
...
             X_train = (
-                self._crf_tokens_to_features(sentence, include_tag_features)
-                for sentence in df_train
+                self._message_to_features(msg, tag_name)
+                for msg in training_data.stream_featurized_messages()
             )
             y_train = (
-                self._crf_tokens_to_tags(sentence, tag_name) for sentence in df_train
+                self._message_to_tags(msg, tag_name)
+                for msg in training_data.stream_featurized_messages()
```
Comment (wochinge) [inline rasa/nlu/extractors/crf_entity_extractor.py:592]: that's potentially dangerous if shuffle would be enabled as X and Y would be shuffled differently, right?

**Review by wochinge — COMMENTED** [review]
How about showing this to Daksh / Vova? I'd be interested in their feedback. I think your prototype seems very promising! 💯

**Comment by twerkmeister** [comment]
I ran some mem profilings on multiwoz dataset with roughly 55k nlu sentences. I compared the current main branch with my branch here. It saves memory all around, but the effect is very much dependent on the specific component. It's the least pronounced for the crf in this case as that model is building up the entire dataset internally itself.

### config 
I think most of it pretty standard, added the crf features here to get `text_dense_features` from the spacy vectors in
```
language: en

pipeline:
  - name: "SpacyNLP"
  - name: "SpacyTokenizer"
  - name: "SpacyFeaturizer"
  - name: "CountVectorsFeaturizer"
  - name: "CRFEntityExtractor"
    max_iterations: 1
    features: [
    ["pos", "pos2"],
    [
      "bias",
      "prefix5",
      "prefix2",
      "suffix5",
      "suffix3",
      "suffix2",
      "pos",
      "pos2",
      "digit",
      "text_dense_features"
    ],
    ["pos", "pos2"],
    ]
    L1_c: 0.01,
    L2_c: 0.05
  - name: "DIETClassifier"
    entity_recognition: False
    epochs: 1
    batch_size: 64
```

### Results
##### main
standard_with_child

##### this branch
streamed_with_child

Memory:
* For each model jumps to about 2.5 GB when loading spacy model and tokenizing all the messages
* During crf data accumulation:
   * main: jumps and hovers at around 5.5 to 6GB
   * this branch: slow increase from 2.5 to about 4.5GB
* During crf training:
   * main: peak memory usage at about 7.5 GB
   * this branch: peak memory usage at about 6 GB
* During Diet training
   * main: constant memory consumption of around 4.3 GB
   * this branch: memory consumption of around 2-2.2 GB

Time:
* main: ~1600s
* this branch: ~1850s   
* Maybe a minute of that goes to CRF training or before that, both of them end at similar times at around 1000s
* The rest seems to be attributable to DIET training
* I think this is mostly expected as 1) the current implementation for the keras sequence is very rough, and 2) caching on disc is fundamentally a trade off between memory and time

File: rasa/nlu/extractors/crf_entity_extractor.py (hunk 4 above)
Comment (twerkmeister) [reply rasa/nlu/extractors/crf_entity_extractor.py:592]: Yes, indeed that would break. The reason this is here is that you can only consume a generator once. However, you could split the generator, if you can guarantee that no copy is advancing more than `n` elements ahead of the others. It would be possible to do this here as the crf internally zips x and y and then runs through them together.

**Comment by twerkmeister** [comment]
> How about showing this to Daksh / Vova? I'd be interested in their feedback. I think your prototype seems very promising! 💯

I'd be interested in their thoughts aswell

**Comment by twerkmeister** [comment]
With splitting (reusing) the generators inside the crf the consumption is again 500mb lower. Timing isn't really comparable as I was doing a lot of other stuff while profiling.

streamed_with_child_combined_crf_stream

File: rasa/nlu/extractors/crf_entity_extractor.py (hunk 4 above)
Comment (twerkmeister) [reply rasa/nlu/extractors/crf_entity_extractor.py]: updated that code again a bit

**Comment by joejuzl** [comment]
How does changing the `MSG_PER_CHUNK` affect the memory usage?

File: rasa/nlu/classifiers/diet_classifier.py (hunk 1 above)
Comment (dakshvar22) [reply rasa/nlu/classifiers/diet_classifier.py:848]: > Yeah maybe something like Dataset.from_generator would be the right way to go

Constructing a `tf.data.Dataset` from `Dataset.generator` becomes cumbersome when you need control on what should happen after the end of every epoch. If you need that control you need to write a custom `train_step` function instead of replying on `keras.model.fit()` to handle that for you. 

[tf.keras.utils.Sequence](https://www.tensorflow.org/api_docs/python/tf/keras/utils/Sequence#on_epoch_end) explicitly has a function for it. That's why we shifted from `tf.data.Dataset.from_generator` to `tf.keras.utils.Sequence` during the transition from [pure TF to reusing Keras methods](https://github.com/RasaHQ/rasa/pull/7438).

However, we recently recognized that tensorflow has [breaking support for `tf.RaggedTensors` inside `tf.keras.utils.Sequence`.](https://github.com/tensorflow/tensorflow/issues/48188#issue-844491927). 
Using `tf.RaggedTensors` will simplify a LOT of code inside our TF models. So, I am all up for exploring how we can use `tf.data.Dataset.from_generator` without having to write our own `model.fit()` function.

File: rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py (hunk 3 above)
Comment (dakshvar22) [reply rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py:716]: I still haven't been able to wrap my head around how it's working, but I okay considering it as some sort of magic 😅  for now which sets the features computed by the specific featurizer for that message, right?

Comment (wochinge) [reply rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py:716]: >  for now which sets the features computed by the specific featurizer for that message, right?

Yes. With the advantage that we control how features are added outside of the components (which in turn means the `Component`s don't have to know about these details). Whether we write the features to disks for chunks or keep them in memory is completely abstracted and exchangable.

Comment (wochinge) [reply rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py:716]: And `Components` are also not aware where data is coming from. It could be in-memory, read from disk or some stuff is coming from in-memory and some is coming from disk.

Comment (dakshvar22) [reply rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py:716]: > With the advantage that we control how features are added outside of the components (which in turn means the Components don't have to know about these details).

But we do make a distinction between different kind of features? for e.g. - sparse and dense features.

**Comment by twerkmeister** [comment]
> How does changing the `MSG_PER_CHUNK` affect the memory usage?

The larger the chunk the more memory is used. At the extreme you could have a chunk that is the size of the dataset even for a big dataset.

The features in this particular setup for 55k messages are about 2GB in memory. So 512 messages are about 20MB.

For the shuffling I am currently using 3 random chunks. So that would make 60mb when using shuffling.

Both the size and number of chunks when shuffling is adjustable.

**Comment by twerkmeister** [comment]
@wochinge @dakshvar22 Brief update on the benchmarking over multiple epochs:

#### Balanced Batching and Preprocessing
Balanced batching can make datasets quite a bit. If you look at my old two graphs of main vs streaming again there's something interesting. Turns out the noisy pattern at the end is actually the DIET training. So it was much faster on the streaming branch, but only because main used balacned batching by default. Given that the nlu data of multiwoz follows a power law for examples per intent (few intents have thousands of examples, and a lot of intents only have a few examples) balanced batching effectively really more than doubled the size of the dataset during training. 

The first long flat stretch in the streaming training on the other hand was a preprocessing step that was very costly on streaming form disc. I would load the data more than a 100 times from disc during this step in the first naive implementation.

With the batching strategy set to `sequence` (no rebalancing) and adjusting the preproccessing slightly, the two systems are now more comparable

#### Timing comparison
I ran the experiments multiple times and consistently observed a 10-15% time penalty for the streaming solution. So for example, on the latest run streaming would take 39s per epoch on an 80% multiwoz train set, and in memory would take 34s. That was consistent over three runs, doing 15-25 epochs each on a google cloud k80 instance.
